File: 2121-intervals-between-identical-elements/2121-intervals-between-identical-elements.js
/**
 * @param {number[]} arr
 * @return {number[]}
 */

var getDistances = function(arr) {
    const buckets = {};
    arr.forEach((el, i) => {
        buckets[el] = buckets[el] || [];
        buckets[el].push(i);
    });
    const ans = new Array(arr.length).fill(0);
    Object.keys(buckets).forEach(el => {
        const indices = buckets[el];
        let [l, r, total] = [0, indices.length - 1, 0];
        indices.forEach(i => total += i - indices[0]);
        ans[indices[0]] = total;
        
        for (let i = 0; i < indices.length - 1; i++) {
            l++;
            const delta = indices[i + 1] - indices[i];
            total += delta * l - delta * r;
            r--;
            ans[indices[i + 1]] = total;
        }
    })
    return ans;
};

File: 2121-intervals-between-identical-elements/2121-intervals-between-identical-elements.rb
# @param {Integer[]} arr
# @return {Integer[]}
def get_distances(arr)
    buckets = Hash.new { |h, k| h[k] = [] }
    arr.each_with_index { |el, i| buckets[el] << i }
    ans = [0] * arr.length
    buckets.keys.each do |el|
        indices = buckets[el]
        l, r, total = 0, indices.length - 1, 0
        indices.each { |i| total += i - indices[0] }
        ans[indices[0]] = total
        
        (1...indices.length).each do |i|
            weight = l - r + 1
            total += weight * (indices[i] - indices[i - 1])
            l += 1
            r -= 1
            ans[indices[i]] = total
        end
    end
    return ans
end